play-json/src/main/scala/com/lightbend/lagom/scaladsl/playjson/PlayJsonSerializer.scala

```diff
@@ -66,16 +67,17 @@ private[lagom] final class PlayJsonSerializer(val system: ExtendedActorSystem, r
   override def fromBinary(bytes: Array[Byte], manifest: String): AnyRef = {
     val startTime = if (isDebugEnabled) System.nanoTime else 0L
 
-    val i = manifest.lastIndexOf('#')
-    val fromVersion = if (i == -1) 1 else manifest.substring(i + 1).toInt
-    val manifestClassName = if (i == -1) manifest else manifest.substring(0, i)
+    val (fromVersion: Int, manifestClassName: String) = parseManifest(manifest)
 
     val migration = migrations.get(manifestClassName)
 
     val migratedManifest = migration match {
-      case Some(migration) if migration.currentVersion > fromVersion =>
+      case Some(migration) if (migration.currentVersion > fromVersion) =>
         migration.transformClassName(fromVersion, manifestClassName)
-      case None => manifestClassName
+      case Some(migration) if (migration.currentVersion < fromVersion) =>
+        throw new IllegalStateException(s"Migration version ${migration.currentVersion} is " +
+          s"behind version $fromVersion of deserialized type [$manifestClassName]")
```

Comment:
Am I correct to say the problem in read-side is still not too bad, because the readside of non-upgraded node will fail - and fail to persist offset - upon seeing a new versioned event. Then it will block the rest of the entity stream on that node as retries continue keeps failing to update offset. However, it is "self-healing" because once its upgraded, then it can continue to process that events, and and the system will still be in a consistent state